azure-pipelines.yml

```diff
@@ -30,22 +32,59 @@ steps:
 
 - script: |
     set -x
-    docker pull $(docker_path):latest || true
-  displayName: "Pull latest docker image to cache"
+    docker pull $(docker_path):$GIT_BRANCH || true
+    docker pull $(docker_path)-middleman:$GIT_BRANCH || true
+  displayName: "Pull any existing images for this branch to cache"
+  # When this branch is first created, there won't be any images to cache, even though master
+  # would probably be an acceptable cache. We could improve build times by pulling master, and
+  # tagging it with the branch, locally.
+  env:
+    GIT_BRANCH: $(Build.SourceBranchName)
 - script: |
     set -x
+    $DOCKER_OVERRIDE build middleman
     $DOCKER_OVERRIDE build
+  displayName: Build Docker Images
+  env:
+    DOCKER_OVERRIDE: $(dockerOverride)
+    dockerHubUsername: $(dockerHubUsername)
+    dockerHubImageName: $(imageName)
+    CC_TEST_REPORTER_ID: $(ccReporterID)
+    GIT_BRANCH: $(Build.SourceBranchName)
+
+- task: Docker@1
+  displayName: Docker Hub login
+  inputs:
+    command: "login"
+    containerregistrytype: Container Registry
+    dockerRegistryEndpoint: DfE Docker Hub
+- task: Docker@1
+  displayName: Push image $(dockerHubUsername)/$(imageName):$(Build.SourceBranchName)
+  condition: and(succeeded(), ne(variables['Build.SourceBranch'], 'refs/heads/master'))
```

Comment:
Ya, I think you're right and `master` should be pushed up too. 👍

Reply:
Oh, wait, I see now. `master` only gets pushed if the tests run without errors. Along with the `$(Build.BuildNumber)` image. See line 144 of this file. The point of pushing the branches up early, before the tests run, is that it could potentially save time on a subsequent build.

Thinking about that, though, I think that that last assumption is risky and wrong: we should probably never push the image before the tests run. First of all, no matter the branch we never want to have an image with failing code pushed up, and secondly I'm not sure we'd be saving a lot of time with caching a broken image anyway.